ral-facilities/datagateway
# Make DataGateway Work with API's ICAT Backend

## Description
In a sentence, these changes allow DataGateway to use DataGateway API's Python ICAT backend, the backend to be used in production. The majority of these changes revolve around moving attribute names from SNAKE_CASE to camelCase, so many of these commits (especially a2c7fde001a4f10456aead012951db029737a495) are the result of find and replace. However, there foreign key attributes have been removed (e.g. `INVESTIGATION_ID` on `Dataset`) as the ICAT backend doesn't make use of foreign keys. 

There's one _very_ minor change in functionality - full names of users are now displayed as per b7f0642a3fa0492ca9f2948690e856351ed824c0.

I've left a couple of TODOs in (search for `TODO -` which will display mine) just to serve as reminders about a couple things. They should probably get removed before getting merged, but I wanted to keep them there until _everything_ passes etc.

On that note, there are still a few outstanding things on this branch:
- My data table views (both ISIS and Diamond) have issues on their e2e tests. There is an issue for this (#659)
- Lots of e2e tests on Search fail due to a slow query involving facilityCycles that are included in said query. Again, there is an issue for this (#658) (EDIT: this is fixed as of 26/05/21)
- On dataview, e2e tests on Diamond proposals (card & table view) fail - this is because ICAT Server 4.11.1 needs to be installed on preprod (an request to the API with a distinct filter on multiple fields which needs the newer version of ICAT Server)
- `cartSelection.spec.ts` fails due to issues with the my data views

The CI will inevitably fail on the e2e tests because the DB backend is on preprod, rather than ICAT backend. I've made some screenshots to show the current state of the tests:

Unit Tests for all components- all pass nicely:
![all-unit-tests](https://user-images.githubusercontent.com/32678030/115035457-81006f00-9ec4-11eb-8dec-f4e85f824ce2.JPG)

DataView Card View e2e - all pass except for Diamond proposals (4.11.1) as above
![dataview-e2e-batch1](https://user-images.githubusercontent.com/32678030/115035821-de94bb80-9ec4-11eb-8826-82ec0b8f29c2.JPG)

DataView Table View e2e - all pass except for Diamond proposals (4.11.1) and mydata (issues above)
![dataview-e2e-batch2](https://user-images.githubusercontent.com/32678030/115035971-fc622080-9ec4-11eb-9296-0aa061926900.JPG)

DataView Remaining e2e - all pass except for cartSelection (mydata)
![dataview-e2e-batch3](https://user-images.githubusercontent.com/32678030/115036137-24ea1a80-9ec5-11eb-959c-ae18be7047b4.JPG)

Download e2e - all pass
![download-e2e](https://user-images.githubusercontent.com/32678030/115036340-56fb7c80-9ec5-11eb-8ac7-e37c43575f24.JPG)

Search e2e - lots of fails due to facilityCycles include (issue above)
![search-e2e](https://user-images.githubusercontent.com/32678030/115036434-6bd81000-9ec5-11eb-96b3-f282a3db4baa.JPG)

## Testing instructions
To test this, the ICAT backend of the API should be used. Specifically, use `temp-distinct-client-handling-merge` branch on DataGateway API - this has a mixture of the new client handling solution and the fix for the distinct filter and is the branch that's most like the API that'll be run in production at time of writing. For client handling configuration, I recommend the following settings for testing. You should be able to get away with `client_pool_init_size: 3` but the below settings are the ones that have actually been tested with this frontend branch:
```json
"client_cache_size": 10,
"client_pool_init_size": 7,
"client_pool_max_size": 10,
```

As mentioned above, ICAT Server 4.11.1 is required for the API from now on.


## Agile board tracking
connect to #94
